src/pke/lib/keyswitch/keyswitch-hybrid.cpp

```diff
@@ -90,31 +90,37 @@ EvalKey<DCRTPoly> KeySwitchHYBRID::KeySwitchGen(const PrivateKey<DCRTPoly> oldKe
     std::vector<DCRTPoly> av(numPartQ);
     std::vector<DCRTPoly> bv(numPartQ);
 
-    std::vector<NativeInteger> PModq                     = cryptoParams->GetPModq();
-    std::vector<std::vector<NativeInteger>> PartQHatModq = cryptoParams->GetPartQHatModq();
+    std::vector<NativeInteger> PModq = cryptoParams->GetPModq();
+    usint numPerPartQ                = cryptoParams->GetNumPerPartQ();
 
     for (usint part = 0; part < numPartQ; part++) {
         DCRTPoly a = DCRTPoly(dug, paramsQP, Format::EVALUATION);
         DCRTPoly e(dgg, paramsQP, Format::EVALUATION);
         DCRTPoly b(paramsQP, Format::EVALUATION, true);
 
-        // The part with basis Q
-        for (usint i = 0; i < sizeQ; i++) {
-            const NativeInteger& qi = paramsQ->GetParams()[i]->GetModulus();
-            auto ai                 = a.GetElementAtIndex(i);
-            auto ei                 = e.GetElementAtIndex(i);
-            auto sNewi              = sNewExt.GetElementAtIndex(i);
-            auto sOldi              = sOld.GetElementAtIndex(i);
-            auto factor             = PModq[i].ModMulFast(PartQHatModq[part][i], qi);
-            b.SetElementAtIndex(i, -ai * sNewi + factor * sOldi + ns * ei);
+        usint startPartIdx = numPerPartQ * part;
+        usint endPartIdx   = (sizeQ > startPartIdx + numPerPartQ) ? (startPartIdx + numPerPartQ) : sizeQ;
+
+        for (usint i = 0; i < startPartIdx; i++) {
```

Comment:
I suggest combining three loops into a single loop. They appear to be the same except where we deal with the actual digit.

Reply:
I also suggest adding a comment that that P*[s_old] is only applied for the current digit to match it to the current ciphertext digit since it is not explicitly stated in any papers.

Reply:
updated